File: forgerock-authenticator/ios/Classes/FRAStorageClient.swift
//
//  Copyright (c) 2022-2026 ForgeRock. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import FRCore
import FRAuthenticator

struct FRAStorageClient: StorageClient {
    
    /// The key identifier for storing device token in Keychain
    let deviceTokenIdentifier = "deviceToken"
    
    /// Keychain Service types for all storages in SDK
    enum KeychainStoreType: String {
        case account = ".account"
        case mechanism = ".mechanism"
        case notification = ".notification"
        case pushDeviceToken = ".pushDeviceToken"
        case backup = ".backup"
    }
    
    var accountStorage: KeychainService
    var mechanismStorage: KeychainService
    var notificationStorage: KeychainService
    var pushDeviceTokenStorage: KeychainService
    var backupStorage: KeychainService
    let keychainServiceIdentifier = "com.forgerock.authenticator.keychainservice.local"
    let notificationsMaxSize = 20
    
    init() {
        self.accountStorage = KeychainService(service: keychainServiceIdentifier + KeychainStoreType.account.rawValue)
        self.mechanismStorage = KeychainService(service: keychainServiceIdentifier + KeychainStoreType.mechanism.rawValue)
        self.notificationStorage = KeychainService(service: keychainServiceIdentifier + KeychainStoreType.notification.rawValue)
        self.pushDeviceTokenStorage = KeychainService(service: keychainServiceIdentifier + KeychainStoreType.pushDeviceToken.rawValue)
        self.backupStorage = KeychainService(service: keychainServiceIdentifier + KeychainStoreType.backup.rawValue)
    }
    
    @discardableResult func setAccount(account: Account) -> Bool {
        do {
            let accountData = try NSKeyedArchiver.archivedData(withRootObject: account, requiringSecureCoding: true)
            return self.accountStorage.set(accountData, key: account.identifier)
        }
        catch {
            FRALog.e("Failed to serialize Account object: \(error.localizedDescription)")
            return false
        }
    }
    
    @discardableResult func removeAccount(account: Account) -> Bool {
        return self.accountStorage.delete(account.identifier)
    }
    
    func getAccount(accountIdentifier: String) -> Account? {
        guard let accountData = self.accountStorage.getData(accountIdentifier) else { return nil }
        if let account = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Account.self, from: accountData) {
            return account
        }
        else {
            return nil
        }
    }
    
    func getAllAccounts() -> [Account] {
        var accounts: [Account] = []
        if let items = self.accountStorage.allItems() {
            for item in items {
                if let accountData = item.value as? Data, let account = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Account.self, from: accountData) {
                    accounts.append(account)
                }
            }
        }
        return accounts.sorted { (lhs, rhs) -> Bool in
            return lhs.timeAdded.timeIntervalSince1970 < rhs.timeAdded.timeIntervalSince1970
        }
    }
    
    @discardableResult func setMechanism(mechanism: Mechanism) -> Bool {
        do {
            let mechanismData = try NSKeyedArchiver.archivedData(withRootObject: mechanism, requiringSecureCoding: true)
            return self.mechanismStorage.set(mechanismData, key: mechanism.identifier)
        }
        catch {
            FRALog.e("Failed to serialize Mechanism object: \(error.localizedDescription)")
            return false
        }
    }
    
    @discardableResult func removeMechanism(mechanism: Mechanism) -> Bool {
        return self.mechanismStorage.delete(mechanism.identifier)
    }

    func getMechanismsForAccount(account: Account) -> [Mechanism] {
        var mechanisms: [Mechanism] = []
        if let items = self.mechanismStorage.allItems() {
            for item in items {
                if let mechanismData = item.value as? Data,
                   let mechanism = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Mechanism.self, from: mechanismData) {
                    if mechanism.issuer == account.issuer && mechanism.accountName == account.accountName {
                        mechanisms.append(mechanism)
                    }
                }
            }
        }
        return mechanisms.sorted { (lhs, rhs) -> Bool in
            return lhs.timeAdded.timeIntervalSince1970 < rhs.timeAdded.timeIntervalSince1970
        }
    }
    
    func getAllMechanisms() -> [Mechanism]? {
        var mechanisms: [Mechanism] = []
        if let items = self.mechanismStorage.allItems() {
            for item in items {
                if let mechanismData = item.value as? Data,
                   let mechanism = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Mechanism.self, from: mechanismData) {
                    mechanisms.append(mechanism)
                }
            }
        }
        return mechanisms
    }
    
    func getMechanism(mechanismIdentifier: String) -> Mechanism? {
        let id = getMechanismId(mechanismId: mechanismIdentifier)
        if let mechanismData = self.mechanismStorage.getData(id),
           let mechanism = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Mechanism.self, from: mechanismData) {
            return mechanism
        }
        else {
            return nil
        }
    }
    
    private func getMechanismId(mechanismId: String) -> String {
        return mechanismId.replacingOccurrences(of: "#", with: "-")
    }
    
    func getMechanismForUUID(uuid: String) -> Mechanism? {
        if let items = self.mechanismStorage.allItems() {
            for item in items {
                if let mechanismData = item.value as? Data,
                   let mechanism = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Mechanism.self, from: mechanismData) {
                    if mechanism.mechanismUUID == uuid {
                        return mechanism
                    }
                }
            }
        }
        return nil
    }
    
    func getAllMechanismsGroupByUID() -> [String: Any]? {
        var mechanismMap: [String: Any] = [:]
        if let items = self.mechanismStorage.allItems() {
            for item in items {
                if let mechanismData = item.value as? Data,
                   let mechanism = try? NSKeyedUnarchiver.unarchivedObject(ofClass: Mechanism.self, from: mechanismData) {
                    mechanismMap[mechanism.mechanismUUID] = MechanismConverter.toJson(mechanism:mechanism)
                }
            }
        }
        return mechanismMap
    }
        
    @discardableResult func setNotification(notification: PushNotification) -> Bool {
        do {
            let notificationData = try NSKeyedArchiver.archivedData(withRootObject: notification, requiringSecureCoding: true)
            return self.notificationStorage.set(notificationData, key: notification.identifier)
        }
        catch {
            FRALog.e("Failed to serialize PushNotification object: \(error.localizedDescription)")
            return false
        }
    }
        
    @discardableResult func removeNotification(notification: PushNotification) -> Bool {
        return self.notificationStorage.delete(notification.identifier)
    }
    
    @discardableResult func removeAllNotifications() -> Bool {
        return self.notificationStorage.deleteAll()
    }
    
    func getNotification(notificationIdentifier: String) -> PushNotification? {
        guard let notificationData = self.notificationStorage.getData(notificationIdentifier) else { return nil }
        if let notification = try? NSKeyedUnarchiver.unarchivedObject(ofClass: PushNotification.self, from: notificationData) {
            return notification
        }
        else {
            return nil
        }
    }
    
    func getAllNotificationsForMechanism(mechanism: Mechanism) -> [PushNotification] {
        var notifications: [PushNotification] = []
        if let items = self.notificationStorage.allItems() {
           for item in items {
               if let notificationData = item.value as? Data,
                  let notification = try? NSKeyedUnarchiver.unarchivedObject(ofClass: PushNotification.self, from: notificationData),
                notification.mechanismUUID == mechanism.mechanismUUID {
                   notifications.append(notification)
               }
           }
        }
        return notifications.sorted { (lhs, rhs) -> Bool in
            return lhs.timeAdded.timeIntervalSince1970 < rhs.timeAdded.timeIntervalSince1970
        }
    }
    
    func getAllNotifications() -> [PushNotification] {
        var notifications: [PushNotification] = []
        if let items = self.notificationStorage.allItems() {
           for item in items {
               if let notificationData = item.value as? Data, let notification = try? NSKeyedUnarchiver.unarchivedObject(ofClass: PushNotification.self, from: notificationData) {
                   notifications.append(notification)
               }
           }
        }
        notifications = notifications.sorted { (lhs, rhs) -> Bool in
            return lhs.timeAdded.timeIntervalSince1970 > rhs.timeAdded.timeIntervalSince1970
        }
        return self.removeOldNotificationEntries(notifications: &notifications)
    }
    
    func getNotificationByMessageId(messageId: String) -> PushNotification? {
        if let items = self.notificationStorage.allItems() {
           for item in items {
               if let notificationData = item.value as? Data,
                  let notification = try? NSKeyedUnarchiver.unarchivedObject(ofClass: PushNotification.self, from: notificationData),
                  notification.messageId == messageId {
                   return notification
               }
           }
        }
        
        return nil
    }
    
    private func removeOldNotificationEntries(notifications: inout [PushNotification])  -> [PushNotification] {
        FRALog.v("Checking old PushNotification entries to remove...")
        var removedEntries = 0
        while (notifications.count > notificationsMaxSize) {
            self.removeNotification(notification: notifications.last!)
            notifications.removeLast()
            removedEntries+=1
        }
        FRALog.v("\(removedEntries) PushNotification entries removed.")
        return notifications
    }
    
    @discardableResult func setPushDeviceToken(pushDeviceToken: PushDeviceToken) -> Bool {
        do {
            let pushDeviceTokenData = try NSKeyedArchiver.archivedData(withRootObject: pushDeviceToken, requiringSecureCoding: true)
            return self.pushDeviceTokenStorage.set(pushDeviceTokenData, key: deviceTokenIdentifier)
        }
        catch {
            FRALog.e("Failed to serialize PushDeviceToken object: \(error.localizedDescription)")
            return false
        }
    }
    
    
    func getPushDeviceToken() -> PushDeviceToken? {
        guard let pushDeviceTokenData = self.pushDeviceTokenStorage.getData(deviceTokenIdentifier) else { return nil }
        if let pushDeviceToken = try? NSKeyedUnarchiver.unarchivedObject(ofClass: PushDeviceToken.self, from: pushDeviceTokenData) {
            return pushDeviceToken
        }
        else {
            return nil
        }
    }
    
    
    @discardableResult func removePushDeviceToken() -> Bool {
        return self.pushDeviceTokenStorage.delete(deviceTokenIdentifier)
    }
    
    @discardableResult func isEmpty() -> Bool {
        return self.notificationStorage.allItems()?.count == 0 && self.mechanismStorage.allItems()?.count == 0 && self.accountStorage.allItems()?.count == 0
    }
    
    func removeAllData() {
        let secItemClasses = [kSecClassGenericPassword,
            kSecClassInternetPassword,
            kSecClassCertificate,
            kSecClassKey,
            kSecClassIdentity]
        for secItemClass in secItemClasses {
            let query: NSDictionary = [
                kSecClass as String: secItemClass,
                kSecAttrSynchronizable as String: kSecAttrSynchronizableAny
            ]
            SecItemDelete(query)
        }
    }
    
    func getBackup(identifier: String) -> String? {
        if let backupData = self.backupStorage.getData(identifier),
           let data = try? NSKeyedUnarchiver.unarchivedObject(ofClass: NSString.self, from: backupData) {
            return data as String
        }
        else {
            return nil
        }
    }
    
    @discardableResult func setBackup(identifier: String, jsonData: String) -> Bool {
        do {
            let backupData = try NSKeyedArchiver.archivedData(withRootObject: jsonData as NSString, requiringSecureCoding: true)
            return self.backupStorage.set(backupData, key: identifier)
        }
        catch {
            FRALog.e("Failed to serialize String object: \(error.localizedDescription)")
            return false
        }
    }
    
}

File: forgerock-authenticator/android/src/main/java/org/forgerock/android/auth/FRAStorageClient.java
/*
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;

/**
 * Data Access Object which implements StorageClient interface and uses SecureSharedPreferences from
 * forgerock-core SDK to store and load Accounts, Mechanisms and Notifications.
 */
class FRAStorageClient implements StorageClient {

    //Alias to store keys
    private static final String FORGEROCK_SHARED_PREFERENCES_KEYS = "com.forgerock.authenticator.KEYS";

    //Settings to store the data
    private static final String FORGEROCK_SHARED_PREFERENCES_DATA_ACCOUNT = "com.forgerock.authenticator.DATA.ACCOUNT";
    private static final String FORGEROCK_SHARED_PREFERENCES_DATA_MECHANISM = "com.forgerock.authenticator.DATA.MECHANISM";
    private static final String FORGEROCK_SHARED_PREFERENCES_DATA_NOTIFICATIONS = "com.forgerock.authenticator.DATA.NOTIFICATIONS";
    private static final String FORGEROCK_SHARED_PREFERENCES_DATA_BACKUP = "com.forgerock.authenticator.DATA.BACKUP";

    //The SharedPreferences to store the data
    private final SharedPreferences accountData;
    private final SharedPreferences mechanismData;
    private final SharedPreferences notificationData;
    private final SharedPreferences backupData;

    private final HashMap<String, Account> accountMap;
    private final HashMap<String, Mechanism> mechanismMap;
    private final HashMap<String, PushNotification> notificationMap;

    private static final String TAG = DefaultStorageClient.class.getSimpleName();
    private static final int NOTIFICATIONS_MAX_SIZE = 20;

    /**
     * Constructor.
     *
     * @param context application context.
     */
    public FRAStorageClient(Context context) {
        this.accountData = new SecuredSharedPreferences(context,
                FORGEROCK_SHARED_PREFERENCES_DATA_ACCOUNT, FORGEROCK_SHARED_PREFERENCES_KEYS);
        this.mechanismData = new SecuredSharedPreferences(context,
                FORGEROCK_SHARED_PREFERENCES_DATA_MECHANISM, FORGEROCK_SHARED_PREFERENCES_KEYS);
        this.notificationData = new SecuredSharedPreferences(context,
                FORGEROCK_SHARED_PREFERENCES_DATA_NOTIFICATIONS, FORGEROCK_SHARED_PREFERENCES_KEYS);
        this.backupData = new SecuredSharedPreferences(context,
                FORGEROCK_SHARED_PREFERENCES_DATA_BACKUP, FORGEROCK_SHARED_PREFERENCES_KEYS);

        this.accountMap = new HashMap<>();
        this.mechanismMap = new HashMap<>();
        this.notificationMap = new HashMap<>();
    }

    @Override
    public Account getAccount(String accountId) {
        if(this.accountMap.containsKey(accountId)) {
            return this.accountMap.get(accountId);
        } else {
            String json = accountData.getString(accountId, "");
            return Account.deserialize(json);
        }
    }

    @Override
    public List<Account> getAllAccounts() {
        List<Account> accountList;

        if(this.accountMap.isEmpty()) {
            accountList = new ArrayList<>();
            Map<String,?> keys = accountData.getAll();
            for(Map.Entry<String,?> entry : keys.entrySet()){
                if(entry.getValue() != null) {
                    Logger.debug(TAG, "Account map values: ",entry.getKey() + ": " + entry.getValue().toString());
                    Account account = Account.deserialize(entry.getValue().toString());
                    if(account != null) {
                        accountList.add(account);
                        this.accountMap.put(account.getId(), account);
                    }
                }
            }
        } else {
            accountList = new ArrayList<>(this.accountMap.values());
        }

        return accountList;
    }

    @Override
    public boolean removeAccount(Account account) {
        boolean success = accountData.edit()
                .remove(account.getId())
                .commit();

        if(success) {
            this.accountMap.remove(account.getId());
        }

        return success;
    }

    @Override
    public boolean setAccount(Account account) {
        String accountJson = account.serialize();

        boolean success = accountData.edit()
                .putString(account.getId(), accountJson)
                .commit();

        if(success) {
            this.accountMap.put(account.getId(), account);
        }

        return success;
    }

    /**
     * Get all mechanisms stored in the system.
     *
     * @return The complete list of mechanisms.
     */
    public List<Mechanism> getAllMechanisms() {
        List<Mechanism> mechanismList;

        if(this.mechanismMap.isEmpty()) {
            mechanismList = new ArrayList<>();
            Map<String,?> keys = mechanismData.getAll();
            for(Map.Entry<String,?> entry : keys.entrySet()){
                if(entry.getValue() != null) {
                    Logger.debug(TAG, "Mechanism map values: ",entry.getKey() + ": " + entry.getValue().toString());
                    String jsonData = entry.getValue().toString();

                    Mechanism mechanism = Mechanism.deserialize(jsonData);
                    if(mechanism != null) {
                        mechanismList.add(mechanism);
                        this.mechanismMap.put(mechanism.getId(), mechanism);
                    }
                }
            }
        } else {
            mechanismList = new ArrayList<>(this.mechanismMap.values());
        }

        return mechanismList;
    }

    @Override
    public List<Mechanism> getMechanismsForAccount(Account account) {
        List<Mechanism> mechanismList = new ArrayList<>();

        List<Mechanism> allMechanisms = this.getAllMechanisms();
        for(Mechanism mechanism : allMechanisms){
            if(mechanism.getIssuer().equals(account.getIssuer()) &&
                    mechanism.getAccountName().equals(account.getAccountName())){
                mechanismList.add(mechanism);
            }
        }

        return mechanismList;
    }

    private String getMechanismId(String mechanismId) {
        mechanismId = mechanismId.replace('#', '-');
        if (mechanismId.contains("-hotp")) {
            return mechanismId.replaceAll("-hotp", "-otpauth");
        } else if (mechanismId.contains("-totp")) {
            return mechanismId.replaceAll("-totp", "-otpauth");
        } else {
            return mechanismId.replaceAll("-push", "-pushauth");
        }
    }

    public Mechanism getMechanism(String mechanismId) {
        String id = getMechanismId(mechanismId);
        if(this.mechanismMap.containsKey(id)) {
            return this.mechanismMap.get(id);
        } else {
            String json = mechanismData.getString(id, "");
            return Mechanism.deserialize(json);
        }
    }

    @Override
    public Mechanism getMechanismByUUID(String mechanismUID) {
        Mechanism mechanism = null;

        List<Mechanism> allMechanisms;
        if(this.mechanismMap.isEmpty()) {
            allMechanisms = this.getAllMechanisms();
        } else {
            allMechanisms = new ArrayList<>(this.mechanismMap.values());
        }

        for(Mechanism mechanismEntry : allMechanisms){
            if (mechanismEntry.getMechanismUID().equals(mechanismUID)) {
                mechanism = mechanismEntry;
                break;
            }
        }

        return mechanism;
    }

    @Override
    public boolean removeMechanism(Mechanism mechanism) {
        boolean success = mechanismData.edit()
                .remove(mechanism.getId())
                .commit();

        if(success) {
            this.mechanismMap.remove(mechanism.getId());
        }

        return success;
    }

    @Override
    public boolean setMechanism(Mechanism mechanism) {
        String mechanismJson = mechanism.serialize();

        boolean success = mechanismData.edit()
                .putString(mechanism.getId(), mechanismJson)
                .commit();

        if(success) {
            this.mechanismMap.put(mechanism.getId(), mechanism);
        }

        return success;
    }

    /**
     * Get all notifications stored in the system.
     *
     * @return The complete list of notifications.
     */
    @Override
    public List<PushNotification> getAllNotifications() {
        List<PushNotification> pushNotificationList;

        if(this.notificationMap.isEmpty()) {
            pushNotificationList = new ArrayList<>();
            Map<String,?> keys = notificationData.getAll();
            for(Map.Entry<String,?> entry : keys.entrySet()){
                if(entry.getValue() != null) {
                    Logger.debug(TAG, "PushNotification map values: ",
                            entry.getKey() + ": " + entry.getValue().toString());
                    PushNotification pushNotification = PushNotification.deserialize(entry.getValue().toString());
                    if(pushNotification != null) {
                        pushNotificationList.add(pushNotification);
                        this.notificationMap.put(pushNotification.getId(), pushNotification);
                    }
                }
            }
        } else {
            pushNotificationList = new ArrayList<>(this.notificationMap.values());
        }

        Collections.sort(pushNotificationList);
        return this.removeOldNotificationEntries(pushNotificationList);
    }

    private List<PushNotification> removeOldNotificationEntries(List<PushNotification> pushNotificationList) {
        Logger.debug(TAG, "Checking old PushNotification entries to remove...");
        int removedEntries = 0;
        while (pushNotificationList.size() > NOTIFICATIONS_MAX_SIZE) {
            int lastElement = pushNotificationList.size() - 1;
            this.removeNotification(pushNotificationList.get(lastElement));
            pushNotificationList.remove(lastElement);
            removedEntries++;
        }
        Logger.debug(TAG, removedEntries + " PushNotification entries removed.");
        return pushNotificationList;
    }

    @Override
    public List<PushNotification> getAllNotificationsForMechanism(Mechanism mechanism) {
        List<PushNotification> pushNotificationList = new ArrayList<>();

        List<PushNotification> allPushNotifications = this.getAllNotifications();
        for(PushNotification pushNotification : allPushNotifications){
            if(pushNotification.getMechanismUID().equals(mechanism.getMechanismUID())){
                pushNotification.setPushMechanism(mechanism);
                pushNotificationList.add(pushNotification);
            }
        }

        return pushNotificationList;
    }

    @Override
    public boolean removeNotification(PushNotification pushNotification) {
        boolean success = notificationData.edit()
                .remove(pushNotification.getId())
                .commit();

        if(success) {
            this.notificationMap.remove(pushNotification.getId());
        }

        return success;
    }

    /**
     * Remove all the stored {@link PushNotification}
     */
    public void removeAllNotifications() {
        notificationData.edit()
                .clear()
                .commit();
        notificationMap.clear();
    }

    /**
     * Get the PushNotification object with its id
     * @param notificationId The PushNotification unique ID
     * @return The PushNotification object.
     */
    public PushNotification getNotification(String notificationId) {
        if(this.notificationMap.containsKey(notificationId)) {
            return this.notificationMap.get(notificationId);
        } else {
            String json = notificationData.getString(notificationId, null);
            return PushNotification.deserialize(json);
        }
    }

    @Override
    public boolean setNotification(@NonNull PushNotification pushNotification) {
        String notificationJson = pushNotification.serialize();

        boolean success = notificationData.edit()
                .putString(pushNotification.getId(), notificationJson)
                .commit();

        if(success) {
            if(this.notificationMap.isEmpty()) {
                this.getAllNotifications();
            }

            this.notificationMap.put(pushNotification.getId(), pushNotification);
        }

        return success;
    }

    @Override
    public boolean isEmpty() {
        return accountData.getAll().isEmpty() &&
                mechanismData.getAll().isEmpty() &&
                notificationData.getAll().isEmpty();
    }

    /**
     * Get the backup data with its id
     * @param id The backup unique ID
     * @return The backup data.
     */
    public String getBackup(String id) {
        return backupData.getString(id, "");
    }

    /**
     * Add or update the backup to the storage system.
     * @param id The identifier of the data.
     * @param jsonData The data as JSON string to store.
     * @return boolean as result of the operation
     */
    public boolean setBackup(String id, String jsonData) {
        return backupData.edit()
                .putString(id, jsonData)
                .commit();
    }

    /**
     * Remove all the stored {@link Account}, {@link Mechanism} and {@link PushNotification}
     */
    @SuppressLint("ApplySharedPref")
    public void removeAll() {
        accountData.edit()
                .clear()
                .commit();
        mechanismData.edit()
                .clear()
                .commit();
        notificationData.edit()
                .clear()
                .commit();
        backupData.edit()
                .clear()
                .commit();
        accountMap.clear();
        mechanismMap.clear();
        notificationMap.clear();
    }

}

